distributed/client.py

```diff
@@ -235,9 +246,42 @@ def __str__(self):
     __repr__ = __str__
 
 
+class FutureState(object):
+    """A Future's internal state.
+
+    This is shared between all Futures with the same key and client.
+    """
+    __slots__ = ('event', 'status', 'type', 'exception', 'traceback')
+
+    def __init__(self, event):
+        self.event = event
+        self.status = 'pending'
+        self.type = None
+
+    def cancel(self):
+        self.status = 'cancelled'
+        self.event.set()
+
+    def finish(self, type=None):
+        self.status = 'finished'
+        self.event.set()
+        if type is not None:
+            self.type = type
+
+    def lose(self):
+        self.status = 'lost'
+        self.event.clear()
+
+    def set_error(self, exception, traceback):
+        self.status = 'error'
+        self.exception = exception
+        self.traceback = traceback
+        self.event.set()
```

Comment:
As I said, i tried to store Future objects directly on the Client, but ended up with nasty GC issues (`__del__` could be called even though the Future was still alive). Hence the two-class approach.
Note there is no inheritance here, so the class hierarchy remains shallow.

Without storing Futures on the Client, I find that it's neater to have a state class than have Client much with raw dictionaries represent the state.